Ported UI LDAP Authentication tests

abalakh opened the pull request:
Depends on SatelliteQE/airgun#278
Test results:
```python
pytest -v tests/foreman/ui_airgun/test_ldap_authentication.py -k 'test_positive_add_katello_role or test_positive_delete_external_roles or test_positive_update_external_user_roles or test_positive_add_admin_role_with_org_loc'
================================ test session starts =================================
platform darwin -- Python 3.6.4, pytest-3.6.1, py-1.6.0, pluggy-0.6.0 -- /Users/andrii/workspace/py3a/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.22.0, services-1.2.1, mock-1.10.0, forked-0.2, env-0.6.2
collecting 8 items                                                                   2019-01-29 15:51:47 - conftest - DEBUG - BZ deselect is disabled in settings

collected 8 items / 4 deselected

tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_katello_role PASSED [ 25%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_delete_external_roles PASSED [ 50%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_user_roles PASSED [ 75%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_admin_role_with_org_loc PASSED [100%]

===================== 4 passed, 4 deselected in 1006.84 seconds ======================
```

abalakh commented on tests/foreman/ui/test_adusergroup.py:
```diff
@@ -128,108 +123,6 @@ def test_positive_add_admin_role(self):
             session.nav.go_to_content_views()
 
-    @tier2
-    def test_positive_add_foreman_role(self):
```
removed. fully covered in `test_positive_update_external_user_roles`

abalakh commented on tests/foreman/ui/test_adusergroup.py:
```diff
@@ -325,256 +218,6 @@ def test_negative_create_external_with_invalid_name(self):
-            2. Assign some roles to UserGroup.
-            3. Create an External AD UserGroup as per the UserGroup name in AD.
-            4. Delete the External AD UserGroup.
-               Note:- Deletion as of sat6.1 is possible only via CLI and not
```
removed. it's 6.5 and deletion is still not there :(

oshtaier replied:
I would keep that stub as that decision is pretty much out of scope for current effort you are working on

abalakh replied:
ok, returned it back

abalakh commented on tests/foreman/ui/test_adusergroup.py:
```diff
@@ -599,72 +242,3 @@ def test_positive_add_external_user(self):
-
-    @stubbed()
-    @tier2
-    def test_positive_add_foreman_role_with_org_loc(self):
```
removed. as of 6.5, specifying org and loc is required, so existing tests cover this scenario.

oshtaier replied:
yes, and it is expected to perform operation defined by roles exactly in the context of location and organization specified in auth source which is a different coverage that you provided above, but very slight one

abalakh replied:
implemented extra test as per discussion (but katello one, foreman one doesn't really make sense as foreman entities don't usually offer org/loc selection and don't tie up to them)

oshtaier replied:
and since when host group is a katello entity?

oshtaier replied:
host entity also provides selection of org and loc, but it is definitely not a katello entity.
probably we need to use whether katello entity or proper words in out statements

abalakh replied:
I've mixed up katello and foreman, my bad. It's actually the opposite - only foreman entities have org/loc tabs, katello entities don't have anything similar. I'll update test header to show it's foreman test, thx for catching.
Regarding katello test - there's no such easy way to verify assigned org/loc. The only idea i had was to indirectly check that by switching org/loc and making sure entity isn't present in different org/loc combination, but you were against that. So unless you've changed your mind i'll probably delete that stub.

oshtaier replied:
if it is a stub and you do not provide coverage for it - then leave it as it is, but close issue as completed for porting effort. Implementation of stubs is only a good will, not a requirement for current effort

abalakh commented on tests/foreman/ui/test_adusergroup.py:
```diff
@@ -599,72 +242,3 @@ def test_positive_add_external_user(self):
-
-    @stubbed()
-    @tier2
-    def test_positive_add_katello_role_with_org_loc(self):
```
>removed. as of 6.5, specifying org and loc is required, so existing tests cover this scenario.

abalakh commented:
Forgot to include one more test.
```python
pytest -v tests/foreman/ui_airgun/test_ldap_authentication.py -k test_positive_update_external_roles
================================ test session starts =================================
platform darwin -- Python 3.6.4, pytest-3.6.1, py-1.6.0, pluggy-0.6.0 -- /Users/andrii/workspace/py3a/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.22.0, services-1.2.1, mock-1.10.0, forked-0.2, env-0.6.2
collecting 9 items                                                                   2019-01-29 16:20:43 - conftest - DEBUG - BZ deselect is disabled in settings

collected 9 items / 8 deselected

tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_roles PASSED [100%]

====================== 1 passed, 8 deselected in 325.09 seconds ======================
```

oshtaier commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+                session.activationkey.search('')
+            except NoSuchElementException:
+                session.browser.refresh()
+        session.organization.select(module_org.name)
```
nitpick: select both module_org and module_loc and operate with any entity in that context

abalakh replied:
I'd rather not - at this point location `location_name` is selected. Since user is admin it doesn't really matter to him in which context to operate with entities, but that `location_name` loc is created for this test specifically, so such scenario is a bit more isolated and there's less chance to affect other tests.
I can change to `module_loc` if you still think that's the better way, but i can't see any win in that.

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+            'usergroup.name': ldap_usergroup_name,
+            'roles.resources.assigned': [foreman_role],
+            'external_groups.name': 'foobargroup',
+            'external_groups.auth_source': 'LDAP-' + auth_source.name,
```
please create a var and use it in the code
```python
auth_source = 'LDAP-' + auth_source.name
```

abalakh replied:
What do you mean? `auth_source.name` isn't static so i can't create some global const. Creating var in every single test is overkill, and creating separate fixture when tests are already overloaded with lots of fixtures doesn't sound any better to me.

ldjebran replied:
do not think writing at many places in the test the expression 'LDAP-' + auth_source.name is a good idea too.

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+        session.usergroup.create({
+            'usergroup.name': ldap_usergroup_name,
+            'roles.resources.assigned': [foreman_role],
+            'external_groups.name': 'foobargroup',
```
please create a var and use it in the code
```python
group_name = 'foobargroup'  # or a randomly generated string via gen_string 
```

oshtaier replied:
it is not a random name =)

ldjebran replied:
but used in different places

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+        ) as ldapsession:
+            ldapsession.location.create({'name': location_name})
+            assert ldapsession.location.search(location_name)[0]['Name'] == location_name
+            current_user = ldapsession.location.read(location_name)['current_user']
```
perhaps read only current_user
```python
current_user = ldapsession.location.read(location_name, 'current_user')['current_user']
```

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+                session.browser.refresh()
+        session.activationkey.create({'name': ak_name})
+        assert session.activationkey.search(ak_name)[0]['Name'] == ak_name
+        current_user = session.activationkey.read(ak_name)['current_user']
```
same here perhaps we should read only the user

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+        )
+        session.usergroup.create({
+            'usergroup.name': ldap_usergroup_name,
+            'roles.resources.assigned': [foreman_role],
```
perhaps we should add this permission to user and not to usergroup

abalakh replied:
No, looks like we both misread test steps - usergroup should have some permissions assigned (foreman permission in this case), then user should be updated with extra permissions (katello permisions in this particular test). Now test steps seem ok to me.

ldjebran commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +311,343 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+    ) as session:
+        session.location.create({'name': location_name})
+        assert session.location.search(location_name)[0]['Name'] == location_name
+        location = session.location.read(location_name)
```
same here we can read only ['primary', 'current_user']

abalakh replied:
user has access only to primary tab, so no real value here, but i can make that more explicitly for better visibility, why not.

abalakh commented:
@oshtaier @ldjebran responded to some comments, addressed others, added 1 new test and few extra tweaks, pls take a look.
Test results:
```python
pytest -v tests/foreman/ui_airgun/test_ldap_authentication.py -k 'test_positive_add_katello_role or test_positive_delete_external_roles or test_positive_update_external_user_roles or test_positive_add_admin_role_with_org_loc or test_positive_update_external_roles or test_positive_add_katello_role_with_org_loc'
=============================== test session starts ===============================
platform darwin -- Python 3.6.4, pytest-4.0.2, py-1.6.0, pluggy-0.8.1 -- /Users/andrii/workspace/py3a/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.26.1, services-1.3.1, mock-1.10.0, forked-0.2, env-0.6.2, cov-2.6.1
collecting ... 2019-02-07 16:58:14 - conftest - DEBUG - BZ deselect is disabled in settings

collected 10 items / 4 deselected

tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_katello_role PASSED [ 16%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_roles PASSED [ 33%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_delete_external_roles PASSED [ 50%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_user_roles PASSED [ 66%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_admin_role_with_org_loc PASSED [ 83%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_katello_role_with_org_loc PASSED [100%]

============== 6 passed, 4 deselected, 2 warnings in 1155.53 seconds ==============
```

oshtaier commented on tests/foreman/ui_airgun/test_ldap_authentication.py:
```diff
@@ -241,3 +313,371 @@ def test_positive_create_with_idm_org_and_loc(session, ipa_data):
+        ) as ldapsession:
+            ldapsession.location.create({'name': location_name})
+            assert ldapsession.location.search(location_name)[0]['Name'] == location_name
+            current_user = ldapsession.location.read(location_name, 'current_user')['current_user']
```
just a question =)
Location entity/view doesn't have widget like `current_user`, so you get that data from common view. Is it possible to read that data omitting navigation to location All->Current views (so basically whole redundant navigation will be skipped)?

abalakh replied:
I'd like to, but we don't have access to views directly from test layer (or rather shouldn't access them). But there won't be a big overkill since previous command was location search, so basically we only do one extra click and then read user.

abalakh commented:
@oshtaier comments addressed.
```python
pytest -v tests/foreman/ui_airgun/test_ldap_authentication.py -k 'test_positive_add_katello_role or test_positive_delete_external_roles or test_positive_update_external_user_roles or test_positive_add_admin_role_with_org_loc or test_positive_update_external_roles or test_positive_add_foreman_role_with_org_loc or test_positive_add_katello_role_with_org'
================================ test session starts ================================
platform darwin -- Python 3.6.4, pytest-4.0.2, py-1.6.0, pluggy-0.8.1 -- /Users/andrii/workspace/py3a/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /Users/andrii/workspace/robottelo, inifile:
plugins: xdist-1.26.1, services-1.3.1, mock-1.10.0, forked-0.2, env-0.6.2, cov-2.6.1
collecting ... 2019-02-12 19:43:19 - conftest - DEBUG - BZ deselect is disabled in settings

collected 11 items / 4 deselected

tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_katello_role PASSED [ 14%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_roles PASSED [ 28%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_delete_external_roles PASSED [ 42%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_update_external_user_roles PASSED [ 57%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_admin_role_with_org_loc PASSED [ 71%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_foreman_role_with_org_loc PASSED [ 85%]
tests/foreman/ui_airgun/test_ldap_authentication.py::test_positive_add_katello_role_with_org PASSED [100%]

================================= warnings summary ==================================

=============== 7 passed, 4 deselected, 2 warnings in 1486.23 seconds ===============
```
